### PR TITLE
ITSGPU: Hybrid track fitting

### PR DESCRIPTION
**Pull request by mconcas**
@davidrohr just for reference.

### CODE REVIEW

**Comment by mconcas** [comment]
Typical warning I get:

```bash
In file included from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TimeFrameGPU.cu:25:
/ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDADef.h:19: warning: "GPUdic" redefined
   19 | #define GPUdic(CUDA, HIP) GPUCA_GPUdic_select_##CUDA()
      |
In file included from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/Common/GPUCommonDef.h:106,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/../GPU/ITStrackingGPU/Array.h:19,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Definitions.h:49,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Constants.h:24,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TimeFrameGPU.cu:15:
/ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/Common/GPUCommonDefAPI.h:180: note: this is the location of the previous definition
  180 | #define GPUdic(...) GPUd()
      |
In file included from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TimeFrameGPU.cu:25:
/ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDADef.h:19: warning: "GPUdic" redefined
   19 | #define GPUdic(CUDA, HIP) GPUCA_GPUdic_select_##CUDA()
      |
In file included from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/Common/GPUCommonDef.h:106,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/../GPU/ITStrackingGPU/Array.h:19,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Definitions.h:49,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Constants.h:24,
                 from /ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TimeFrameGPU.cu:15:
/ssd_data/builds/gpu/sw/SOURCES/O2/hybrid-gpuframework-based/0/GPU/Common/GPUCommonDefAPI.h:180: note: this is the location of the previous definition
  180 | #define GPUdic(...) GPUd()
      |
```

**Comment by davidrohr** [comment]
The CMake changes look good.
For the memory allocation. I am not 100% convinced that I like making GPUTracking a dependency for ITS tracking. I think it should rather be the other way around. E.g. the GPU CUDA and HIP libraries create the traits.

If I understand correctly, you need it only for the memory allocation. Instead of having a pointer to the ITS chain in the ITS tracking code, you could just create a lambda for the memory allocation, and store a pointer to that in the ITS tracking code.

**Comment by mconcas** [comment]
> The CMake changes look good. For the memory allocation. I am not 100% convinced that I like making GPUTracking a dependency for ITS tracking. I think it should rather be the other way around. E.g. the GPU CUDA and HIP libraries create the traits.
> 
> If I understand correctly, you need it only for the memory allocation. Instead of having a pointer to the ITS chain in the ITS tracking code, you could just create a lambda for the memory allocation, and store a pointer to that in the ITS tracking code.

I now removed the `O2::GPUTracking` dependency by having an `ExternalAllocator` class that is then specialised in the ITSChain. Seems much better approach, I could not figure out something better than this.

**Comment by mconcas** [comment]
Currently stuck because of some badly ordered/conflicting `GPUheader***` inclusion.
If I just add the `Propagator.cxx` it complains of some undefined macros.

```bash
[...]
Detectors/Base/src/Propagator.cxx(697): error: identifier "gGPUConstantMemBuffer" is undefined
      const auto* f = &(gGPUConstantMemBuffer.v).param.polynomialField;
```
Adding `#include "GPUConstantMem.h"` when `O2::GPUTracking` was linked made it complain about the redefinition of other macros and broke the compilation.
Now, adding the same header just complains about not finding all the other included headers, of course.
I tried adding some of them manually in the `PRIVATE_INCLUDE_DIR`, but then I realised that I was pulling in so much unwanted (?) stuff such as TRD and TPC-specific data formats headers, their reconstruction headers, etc.
Need to review this part.

**Comment by davidrohr** [comment]
The problem is actually more fundamental.
That is a gpu constant memory buffer that holda pointers and constants.
If you just include it in what you compile, GPUReconstruction will not initialize it properly, since it doesn't know about it :(.
So the Propagator will actually not work.

We really need a better way of doing this, but that means also restructurimg a large amount of the existing GPUCode.
The only solution I see for now is to make libGPUTracking compile all GPU code that uses any of the common gpu framework stuff.

But I'll think about it a bit. Let's talk about it next week. Monday is a bit difficult for me. How about Tuesday after the RC meeting or after lunch?

Kind Regards
David Rohr

**Comment by mconcas** [comment]
> The problem is actually more fundamental. That is a gpu constant memory buffer that holda pointers and constants. If you just include it in what you compile, GPUReconstruction will not initialize it properly, since it doesn't know about it :(. So the Propagator will actually not work. 
We really need a better way of doing this, but that means also restructuring a large amount of the existing GPUCode. The only solution I see for now is to make libGPUTracking compile all GPU code that uses any of the common gpu framework stuff. 

Since the Propagator is anyways taken from the GPU framework, 
https://github.com/AliceO2Group/AliceO2/blob/89d7ced9dc6305575bf5584976dbac10ba7ee94e/GPU/GPUTracking/Global/GPUChainITS.cxx#L95
I would expect it to be properly initalised. If I make for now my compilation to include all the required headers, would it potentially work or am I missing something?
If so, I would for the time being focus on the definitions of the macros that make the compilation fail (something is just wrong with the way I try to fix it, probably) then I could have the fitting working as a standalone component. 

> But I'll think about it a bit. Let's talk about it next week. Monday is a bit difficult for me. How about Tuesday after the RC meeting or after lunch?

Tuesday is fine, we have ITS software meeting at 14pm (~45 min usually), a part from that I can meet any time.

**Comment by davidrohr** [comment]
No, it will not work even if you include all headers. Since you will get a second instance of the constant memory, which will not be initialized.

**Comment by mconcas** [comment]
> No, it will not work even if you include all headers. Since you will get a second instance of the constant memory, which will not be initialized.

Why a second instance?
Didn't we use  `mychain->processorsShadow()->calibObjects.o2Propagator`precisely to get the pointer to the allocated constmem?🫤

**Comment by davidrohr** [comment]
> > No, it will not work even if you include all headers. Since you will get a second instance of the constant memory, which will not be initialized.
> 
> Why a second instance? Didn't we use `mychain->processorsShadow()->calibObjects.o2Propagator`precisely to get the pointer to the allocated constmem?🫤

well, that is just how CUDA works. If we have 2 compilation units in 2 different libraries, then both will get the constant memory symbol, both accessing only its local instance. So GPUReconstruction will initialize its own, while kernels in your library will use the other uninitialized ones.

Not sure if this could be solved with CUDA dynamic linking. Need to read up about this...

**Comment by davidrohr** [comment]
> Didn't we use `mychain->processorsShadow()->calibObjects.o2Propagator`precisely to get the pointer to the allocated constmem?🫤

There is no pointer to constmem, at least not on the device side. Const mem is resolved by the const symbol. If you check the CUDA assembler code, you'll see that CUDA accesses constant memory differently than global memory.

In that case, calibObjects is indeed in constantMemory, and I have a pointer to the propagator in there. But the propagator itself is not in constant memory, I beliefe, but the parameters are.

**Comment by mconcas** [comment]
@davidrohr, this PR can now be reviewed. HIP still needs to be tested (I am waiting for my turn on EPNs).

Cosmetics cleanup can happen later, before merging.

File: dependencies/FindO2GPU.cmake
Comment (davidrohr) [inline dependencies/FindO2GPU.cmake]: Would be interesting to check if `--generate-line-info` can affect the performance of the code. I believe not, but should be tested.

**Review by davidrohr — DISMISSED** [review]
Fine with me, let's wait for the CI.

File: dependencies/FindO2GPU.cmake
Comment (mconcas) [reply dependencies/FindO2GPU.cmake]: Ah, indeed, this is a leftover of the debugging. I will remove it for the time being.

File: GPU/GPUTracking/Global/GPUChain.h
Comment (davidrohr) [inline GPU/GPUTracking/Global/GPUChain.h]: Why do you need this function to be public? As far as I can tell, it is only used by GPUChainITS, which derives from GPUChain.

In general, if needed, I have no problem with makeing it public, but then I'd actually make the whole block of functions public.

Comment (mconcas) [reply GPU/GPUTracking/Global/GPUChain.h]: You are right, must be a remnant of some kind of test of mine.

**Comment by davidrohr** [comment]
Do you already want to squash the commits in the way you want? Or shall we squash-merge the commit when the CI is green?

**Comment by mconcas** [comment]
> Do you already want to squash the commits in the way you want? Or shall we squash-merge the commit when the CI is green?

I am now also reverting a change to `TrackParametrizationWithError::getChi2(***)` method. This apparently silently breaks at runtime. It essentially does not produce the required symbol but does not complain if in incremental build. 
Fortunately macos CI finds it, because on ubuntu and EPNs this was not spotted.

Later I'll resquash to relevant commits

**Comment by mconcas** [comment]
Ok, HIP tested on EPN. It works (10% faster with same blocks/threads). Results seem consistent.
red: AMD blue Nvidia.

Screenshot 2023-10-27 at 19 19 59
Screenshot 2023-10-27 at 19 13 37

**Comment by mconcas** [comment]
@davidrohr From now on, to me, this can be squashed and merged anytime.

**Pull request merged**
